File: lib/features/feed/domain/entities/post.dart
enum PostType { spark, voice, moment, vibe, soundDrop, moodBoard }

class Post {
  const Post({
    required this.id,
    required this.authorId,
    required this.authorName,
    this.authorUsername,
    this.authorAvatarUrl,
    required this.circleId,
    required this.circleName,
    required this.type,
    required this.createdAt,
    this.text,
    this.imageUrl,
    this.audioUrl,
    this.waveformData,
    this.audioDurationSeconds,
    this.gradientColors,
    this.momentEmoji,
    this.vibeKey,
    this.reactions = const {},
    this.commentCount = 0,
    this.isReacted = false,
    this.myReaction,
  });

  final String id;
  final String authorId;
  final String authorName;
  final String? authorUsername;
  final String? authorAvatarUrl;
  final String circleId;
  final String circleName;
  final PostType type;
  final DateTime createdAt;

  // Spark
  final String? text;
  final String? imageUrl;

  // Voice
  final String? audioUrl;
  final List<double>? waveformData;
  final int? audioDurationSeconds;

  // Moment
  final List<String>? gradientColors;
  final String? momentEmoji;

  // Vibe
  final String? vibeKey;

  // Reactions
  final Map<String, int> reactions; // emoji -> count
  final int commentCount;
  final bool isReacted;
  final String? myReaction;

  Post copyWith({
    Map<String, int>? reactions,
    int? commentCount,
    bool? isReacted,
    String? myReaction,
  }) {
    return Post(
      id: id,
      authorId: authorId,
      authorName: authorName,
      authorUsername: authorUsername,
      authorAvatarUrl: authorAvatarUrl,
      circleId: circleId,
      circleName: circleName,
      type: type,
      createdAt: createdAt,
      text: text,
      imageUrl: imageUrl,
      audioUrl: audioUrl,
      waveformData: waveformData,
      audioDurationSeconds: audioDurationSeconds,
      gradientColors: gradientColors,
      momentEmoji: momentEmoji,
      vibeKey: vibeKey,
      reactions: reactions ?? this.reactions,
      commentCount: commentCount ?? this.commentCount,
      isReacted: isReacted ?? this.isReacted,
      myReaction: myReaction ?? this.myReaction,
    );
  }
}

File: backend/src/models/Post.js
'use strict';

const mongoose = require('mongoose');

const postSchema = new mongoose.Schema({
  author: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true, index: true },
  type: {
    type: String,
    enum: ['spark', 'voice', 'moment', 'vibe', 'soundDrop', 'moodBoard'],
    required: true,
  },

  // Spark (text)
  text: { type: String, maxlength: 500 },

  // Voice
  audioUrl: String,
  audioDurationSec: Number,
  waveformData: [Number],

  // Moment (image/video)
  mediaUrl: String,
  mediaType: { type: String, enum: ['image', 'video'] },
  thumbnailUrl: String,

  // Vibe check
  vibeTag: String,
  vibeColor: String,

  // Sound Drop
  soundTitle: String,
  soundArtist: String,
  soundPreviewUrl: String,

  // Circles
  circleId: { type: mongoose.Schema.Types.ObjectId, ref: 'Circle', index: true },
  isPublic: { type: Boolean, default: true },

  // Engagement
  reactionsCount: { type: Number, default: 0 },
  commentsCount: { type: Number, default: 0 },
  repostsCount: { type: Number, default: 0 },
  tipsTotal: { type: Number, default: 0 },

  // Sticker reactions map: { stickerId -> count }
  stickerReactions: { type: Map, of: Number, default: {} },

  // Moderation
  isDeleted: { type: Boolean, default: false },
  isHidden: { type: Boolean, default: false },
}, {
  timestamps: true,
});

postSchema.index({ createdAt: -1 });
postSchema.index({ author: 1, createdAt: -1 });
postSchema.index({ circleId: 1, createdAt: -1 });

module.exports = mongoose.model('Post', postSchema);
